From 888c569969726c341f0da0504fc1133d2dc9bf20 Mon Sep 17 00:00:00 2001
From: Yan Markman <ymarkman@marvell.com>
Date: Wed, 29 Nov 2017 19:11:25 +0200
Subject: [PATCH 2197/2241] net: pp2x: fix tx degradation by reduced
 packet-coalescing

PROBLEM:
On 10G-port the iperf-TCP-client (TX) bandwidth occasionally fails
from 9G down to 0.5/1M and stays on that for dozens of minutes.
At that time Driver is working - other TX/RX sessions are ok.
Iperf-client restart begins with 9G.
Time to failure is vary from some seconds to hours.

FINDINGS:
The problem is related to TCP-stack and IPERF-application behavior:
Iperf-TCP is using SKB-CLONE = reuse buffers returned by TX-done.
The time to return SKB back to IPERF is critical especially on 100%
usage of 10G bandwidth.
This time is depending upon Coalescing (tx-time and tx-frames).
Reduce tx-frames from 64 down to 16 eliminates the problem.

STRESS TEST (forcing):
Increase coalescing tx-frames=128 by command
 ethtool -C ethN tx-frames 128
causes for problem in some seconds seconds.

SOLUTION:
Reduce tx-frames=16 causes for more IRQs and increases CPU loading
for about 0.7%.
As Stability-vs-performance tradeoff use default value =32

Change-Id: I98e75737df81eee9d9574ad43bc9557532c982d3
Signed-off-by: Yan Markman <ymarkman@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/47066
Tested-by: iSoC Platform CI <ykjenk@marvell.com>
Reviewed-by: Omri Itach <omrii@marvell.com>
---
 drivers/net/ethernet/marvell/mvpp2x/mv_pp2x.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x.h b/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x.h
index 65d3564..22ce6d9 100644
--- a/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x.h
+++ b/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x.h
@@ -156,7 +156,7 @@
 #define MVPP2_MAX_SHARED	1
 
 /* Coalescing */
-#define MVPP2_TXDONE_COAL_PKTS		64
+#define MVPP2_TXDONE_COAL_PKTS		32
 #define MVPP2_TXDONE_HRTIMER_PERIOD_NS	1000000UL
 #define MVPP2_TX_HRTIMER_PERIOD_NS	50000UL
 #define MVPP2_TXDONE_COAL_USEC		1000
-- 
2.7.4

